src/main/java/org/opensearch/security/dlic/rest/api/AccountApiAction.java

```diff
@@ -183,47 +237,106 @@ protected void handleGet(RestChannel channel, RestRequest request, Client client
+        // bad case: no valid saved_tenant body nor password body
+        if (invalidSavedTenantBody && invalidPasswordBody){
+            badRequestResponse(channel, "Invalid request. Needs a saved_tenant AND/OR (saved_password AND (password OR hash)");
             return;
         }
 
-        internalUserEntry.setHash(hash);
+        // process PUT for 'saved_tenant'
+        if (content.get("saved_tenant") != null){
+            final String newSavedTenant = content.get("saved_tenant").asText();
+            // invalid action; user is not InternalUserV7
+            if (!(internalUser.getCEntry(user.getName()) instanceof InternalUserV7)){
+                badRequestResponse(channel, "Cannot modify saved_tenant for non-InternalUserV7.");
+                return;
+            }
+            InternalUserV7 targetUser = (InternalUserV7) internalUser.getCEntry(user.getName());
+            // each user always has access to the global tenant and their own private tenant
+            if (!(newSavedTenant.equals(DEFAULT_TENANT) || newSavedTenant.equals(PRIVATE_TENANT))){
+                SecurityDynamicConfiguration<?> rolesMappingsConfiguration = load(CType.ROLESMAPPING, false);
+                Set<String> userRoles = new HashSet<>();
+                Set<String> accessibleTenants = new HashSet<>();
+
+                // find user roles
+                for (String role : rolesMappingsConfiguration.getCEntries().keySet()){
+                    RoleMappings rm = (RoleMappings) rolesMappingsConfiguration.getCEntry(role);
+                    if (rm.getUsers().contains(user.getName())){
+                        userRoles.add(role);
+                    }
+                }
 
+                // find accessible tenants based on roles
+                final SecurityDynamicConfiguration<?> rolesConfiguration = load(CType.ROLES, false);
+                for (String roleName : userRoles){
+                    // only using OS 1.0+, which uses V7
+                    // not supported for V6 (pre-OS 1.0)
+                    if (rolesConfiguration.exists((roleName)) && rolesConfiguration.getCEntry(roleName) instanceof RoleV7){
+                        RoleV7 role = (RoleV7) rolesConfiguration.getCEntry(roleName);
+                        accessibleTenants.addAll(role.accessibleTenants());
+                    }
+                }
+
+                // equested tenant is not accessible by target user
+                if (!accessibleTenants.contains(newSavedTenant)){
+                    badRequestResponse(channel, "Target user does not have access to specified tenant");
+                    return;
+                }
+            }
+            targetUser.setSaved_tenant(newSavedTenant);
+        }
+        
+        // process PUT for 'password'/'hash'
+        if (content.get("current_password") != null){
```

Comment:
This should be `if` and `else if` rather than two different `if`

With your way, api can modify both saved tenant and password

Reply:
I remember Vlad suggested to use different class for this API. 
Did you consider that? In that case saved tenant will be different from changing the password.

Reply:
I did; we decided that it made more sense to allow users the freedom to be able to change `saved_tenant`, `password`, or both with one call.

Reply:
Thanks for the explanation